#IRP-293 Discount

andreyvorobyov2 opened the pull request.

Commit:
#IRP-293

Commit:
#IRP-293

Commit:
Merge branch 'develop' into IRP-293

Commit:
#IRP-293

DitriXNew commented on IRP/src/CommonModules/ModelClientServer_V2/Module.bsl:
```diff
@@ -3041,8 +2969,16 @@ Function CalculationsExecute(Options) Export
+	// Manual offers
+	If Options.CalculateManualOffers.Enable Then
+		If Options.OffersOptions.ManualOfferType = PredefinedValue("Enum.ManualOfferTypes.Percent") Then
+			Result.ManualOfferAmount = (Options.QuantityOptions.Quantity * Options.PriceOptions.Price) / 100 * Result.ManualOfferPercent;
```
Надо все эти формулы из всех мест свести в одну точку, щас эта формула встречается много раз

andreyvorobyov2 replied:
а где еще она встречается? это единственное место

DitriXNew commented on IRP/src/CommonModules/ModelClientServer_V2/Module.bsl:
```diff
@@ -3123,7 +3059,8 @@ Function CalculationsExecute(Options) Export
 			If Options.CalculatePriceByTotalAmount.Enable And IsCalculatedRow Then
 				Result.Price = ?(Options.PriceOptions.Quantity = 0, 0, 
-					(Result.TotalAmount / Options.PriceOptions.Quantity) + Result.OffersAmount / Options.PriceOptions.Quantity);  
+					(Result.TotalAmount / Options.PriceOptions.Quantity) 
+					+ (Result.OffersAmount + Result.ManualOfferAmount) / Options.PriceOptions.Quantity);  
```
Вынести

DitriXNew commented on IRP/src/CommonModules/ModelClientServer_V2/Module.bsl:
```diff
@@ -3160,7 +3097,8 @@ Function CalculationsExecute(Options) Export
 			If Options.CalculatePriceByTotalAmount.Enable And IsCalculatedRow Then
 				Result.Price = ?(Options.PriceOptions.Quantity = 0, 0, 
-				((Result.TotalAmount - Result.TaxAmount) / Options.PriceOptions.Quantity)  + Result.OffersAmount / Options.PriceOptions.Quantity);
+				((Result.TotalAmount - Result.TaxAmount) / Options.PriceOptions.Quantity)  
+				+ (Result.OffersAmount + Result.ManualOfferAmount) / Options.PriceOptions.Quantity);
```
Вынести

DitriXNew commented on IRP/src/CommonModules/ModelClientServer_V2/Module.bsl:
```diff
@@ -3198,7 +3136,8 @@ Function CalculationsExecute(Options) Export
 		If Options.CalculatePriceByTotalAmount.Enable And IsCalculatedRow Then
 			Result.Price = ?(Options.PriceOptions.Quantity = 0, 0, 
-			((Result.TotalAmount - Result.TaxAmount) / Options.PriceOptions.Quantity) + Result.OffersAmount / Options.PriceOptions.Quantity);
+			((Result.TotalAmount - Result.TaxAmount) / Options.PriceOptions.Quantity) 
+			+ (Result.OffersAmount + Result.ManualOfferAmount) / Options.PriceOptions.Quantity);
```
Вынести

DitriXNew commented on IRP/src/Documents/PurchaseInvoice/Forms/DocumentForm/Module.bsl:
```diff
@@ -140,8 +140,40 @@ Procedure SetVisibilityAvailability(Object, Form)
+			EndIf;
+		EndDo;
+	EndDo;
+	Return ArrayOfKeys;
```
Вынести в общий модуль

DitriXNew commented on IRP/src/Documents/SalesInvoice/Forms/DocumentForm/Module.bsl:
```diff
@@ -138,8 +138,40 @@ Procedure SetVisibilityAvailability(Object, Form)
+Function GetExternalOffers(val _Object)
+	ArrayOfKeys = New Array();
+	For Each Row_SpecialOffer In _Object.SpecialOffers Do               
+		Rows_RowIDInfo = _Object.RowIDInfo.FindRows(New Structure("Key", Row_SpecialOffer.Key));
+		For Each Row_RowIDInfo In Rows_RowIDInfo Do
+			If ValueIsFilled(Row_RowIDInfo.BasisKey)
+				And ValueIsFilled(Row_RowIDInfo.Basis)
+				And (TypeOf(Row_RowIDInfo.Basis) = Type("DocumentRef.SalesOrder") 
+					Or TypeOf(Row_RowIDInfo.Basis) = Type("DocumentRef.PurchaseOrder")) Then
+				OrderRows = Row_RowIDInfo.Basis.SpecialOffers.FindRows(New Structure("Key", Row_RowIDInfo.BasisKey));
+				For Each OrderRow In OrderRows Do
+					If OrderRow.Offer = Row_SpecialOffer.Offer And ArrayOfKeys.Find(Row_SpecialOffer.Key) = Undefined Then
+						ArrayOfKeys.Add(Row_SpecialOffer.Key);
+					EndIf;
+				EndDo; 
+			EndIf;
+		EndDo;
```
Тоже вынести

andreyvorobyov2 replied:
зачем? что бы нырять между модулями и процедурами теряя при этом ход логики и забивая голову бессмысленными перепрыгиваниями между модулями и процедурами. такой код будет читать невозможно.

Commit:
#IRP-293

Commit:
#IRP-293

Commit:
Merge branch 'develop' into IRP-293

Severnity committed:
Merge branch 'develop' into IRP-293

Severnity committed:
Add tests for manual and selected row discounts

Introduces scenarios to verify document discounts applied to selected rows in Sales Order, Purchase Order, Retail Sales Receipt, and Sales Return documents. Also adds a comprehensive test for manual discount entry and propagation across SO, SI, and SR documents.

Severnity committed:
Update ImportData.feature
